sentry/src/main/java/io/sentry/SentryOptions.java

```diff
@@ -1,60 +1,311 @@
+     */
+    public SentryOptions with(Lookup newLookup) {
+        return from(requireNonNull(newLookup), getDsn(), null);
+    }
+
+    /**
+     * Returns a copy of this options instance with the client factory set to the provided value.
+     *
+     * @param newClientFactory the client factory to use in the new options instance
+     * @return the new options instance
+     */
+    public SentryOptions with(SentryClientFactory newClientFactory) {
+        return from(getLookup(), getDsn(), requireNonNull(newClientFactory));
+    }
 
     /**
-     * Gets the optionally set {@Link SentryClientFactory}.
-     * @return {@link SentryClientFactory}
+     * Gets the value of the configuration key using the {@link #getLookup() lookup} and {@link #getDsn() DSN}.
+     *
+     * @param key the name of the configuration key
+     * @return the value of the configuration key or null if not found
      */
-    public SentryClientFactory getSentryClientFactory() {
-        return sentryClientFactory;
+    @Nullable
+    public String getConfigurationKey(String key) {
+        return lookup.get(key, dsn);
     }
 
     /**
-     * Sets the {@link SentryClientFactory} to be used when initializing the SDK.
-     * @param sentryClientFactory Factory used to create a {@link SentryClient}.
+     * Gets the lookup instance used by this options instance.
+     *
+     * @return the configured lookup instance
      */
-    public void setSentryClientFactory(SentryClientFactory sentryClientFactory) {
-        this.sentryClientFactory = sentryClientFactory;
+    public Lookup getLookup() {
+        return lookup;
     }
 
     /**
-     * Gets the DSN. If not DSN was set, the SDK will attempt to find one in the environment.
-     * @return Data Source Name.
+     * Gets the DSN instance used by this options instance.
+     *
+     * @return the configured DSN instance
      */
-    public String getDsn() {
+    public Dsn getDsn() {
         return dsn;
     }
 
     /**
-     * Sets the DSN to be used by the {@link SentryClient}.
-     * @param dsn Sentry Data Source Name.
+     * Gets the Sentry client factory instance used by this options instance.
+     * @return the configured client factory instance
      */
-    public void setDsn(String dsn) {
-        this.dsn = dsn;
+    public SentryClientFactory getClientFactory() {
+        return clientFactory;
     }
 
     /**
-     * Gets the {@link ResourceLoader} to be used when looking for properties.
-     * @return {@link ResourceLoader}
+     * Returns a new Sentry client obtained from the {@link #getClientFactory() factory}.
+     *
+     * @return the new sentry client or null if the client factory is invalid
      */
-    public ResourceLoader getResourceLoader() {
-        return resourceLoader;
+    @Nullable
+    public SentryClient createClient() {
+        return getClientFactory().createSentryClient(getDsn());
+    }
+
+    private static Dsn resolveDsn(Lookup lookup, @Nullable String dsn) {
```

Comment:
Does this belong in `SentryOptions`? In other SDKs we have it more as a POJO type of class.
Perhaps these methods could be pulled out into an internal helper class.